nvflare/app_common/executors/ham.py

```diff
@@ -0,0 +1,241 @@
+
+    def _pending_clients_status(self):
+        with self._status_lock:
+            if not self.pending_clients:
+                return 0, 0
+
+            received = 0
+            for received_time in self.pending_clients.values():
+                if received_time:
+                    received += 1
+
+            return received, len(self.pending_clients)
+
+    def _update_client_status(self, client_name, status):
+        with self._status_lock:
+            self.pending_clients[client_name] = status
+
+    def execute(self, task_name: str, shareable: Shareable, fl_ctx: FLContext, abort_signal: Signal) -> Shareable:
+        """Execute the assigned task.
+        If we are a leaf node in client hierarchy, we'll execute the task by using the configured executor for the
+        task name "exec_<task_name>". This way different tasks can be handled by different executors.
+
+        If we are not leaf node, we'll wait for results from child clients and then aggregate their results using
+        the configured aggregator.
+
+        Args:
+            task_name: name of the assigned task
+            shareable: task data
+            fl_ctx: FLContext object
+            abort_signal: signal to notify abort
+
+        Returns: task result
+
+        """
+        is_leaf = fl_ctx.get_prop(ReservedKey.IS_LEAF)
+        if is_leaf:
+            return self.learner.execute(task_name, shareable, fl_ctx, abort_signal)
+
+        self.log_info(fl_ctx, "waiting for results from children ...")
+        self.current_round = shareable.get_header(AppConstants.CURRENT_ROUND)
+        self.log_debug(fl_ctx, f"got current_round: {self.current_round}")
+        self.pending_task_id = shareable.get_header(ReservedKey.TASK_ID)
+
+        # Set header to indicate that we are ready to manage child clients
+        # Note: when a child comes to pull task, the communicator only sends it after the task is ready.
+        # This is to avoid the potential race condition that the client gets the task and then quickly submits
+        # result before we are even ready.
+        shareable.set_header(ReservedKey.TASK_IS_READY, True)
+        result = self._do_execute(fl_ctx, abort_signal)
+
+        # reset state
+        self.pending_task_id = None
+        self.pending_clients = {}
+        self.aggregator.reset(fl_ctx)
+        self._process_error = False
+        return result
+
+    def _do_execute(self, fl_ctx: FLContext, abort_signal: Signal) -> Shareable:
+        start_time = time.time()
+        min_received_time = None
+        while True:
+            if abort_signal.triggered:
+                return make_reply(ReturnCode.TASK_ABORTED)
+
+            if self._process_error:
+                # we bail out when any processing error encountered
+                break
+
+            current_time = time.time()
+            if current_time - start_time > self.aggr_timeout:
+                # we have waited long enough
```

Comment:
Should log an error here for timing out.

Reply:
This may or may not be error - some results may be received. Error is logged if nothing is received after the loop.